File: sdks/ts/src/clients/obligations/erc1155/barterUtils.ts
import { abi as atomicPaymentUtilsAbi } from "../../../contracts/utils/AtomicPaymentUtils";
import { getAttestedEventFromTxHash, type ViemClient, writeContract } from "../../../utils";
import type { Erc1155Addresses } from "./index";

export type Erc1155BarterUtilsClient = ReturnType<typeof makeErc1155BarterUtilsClient>;

export const makeErc1155BarterUtilsClient = (viemClient: ViemClient, addresses: Erc1155Addresses) => {
  return {
    address: addresses.barterUtils,

    payErc1155AndCollect: async (escrowUid: `0x${string}`) => {
      const hash = await writeContract(viemClient, {
        address: addresses.barterUtils,
        abi: atomicPaymentUtilsAbi.abi,
        functionName: "payErc1155AndCollect",
        args: [escrowUid],
      });

      const attested = await getAttestedEventFromTxHash(viemClient, hash);
      return { hash, attested };
    },
  };
};

File: sdks/rs/src/clients/obligations/erc1155/barter_utils.rs
//! ERC1155 atomic payment utilities.

use alloy::primitives::FixedBytes;
use alloy::rpc::types::TransactionReceipt;

use crate::contracts;

use super::Erc1155Module;

pub struct BarterUtils<'a> {
    module: &'a Erc1155Module,
}

impl<'a> BarterUtils<'a> {
    pub fn new(module: &'a Erc1155Module) -> Self {
        Self { module }
    }

    pub async fn pay_erc1155_and_collect(
        &self,
        escrow_uid: FixedBytes<32>,
    ) -> eyre::Result<TransactionReceipt> {
        let utility = contracts::utils::AtomicPaymentUtils::new(
            self.module.addresses.barter_utils,
            &self.module.wallet_provider,
        );

        Ok(utility
            .payErc1155AndCollect(escrow_uid)
            .send()
            .await?
            .get_receipt()
            .await?)
    }
}
